### ISSUE TITLE
Large matrices should be passed by address

### ISSUE THREAD

**Issue by yurivict** [mlpack/mlpack#3489]
#### Issue description

For example, [here](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/ffn_impl.hpp#L150).

I stepped through the above Train call and found that the Armadillo arrays "predictors" and "responses" are copied.

```arma::Mat<eT>::Mat``` copies arrays like this:
> arrayops::copy( memptr(), in_mat.mem, in_mat.n_elem );

This code should probably look like this:
```
146│ typename MatType::elem_type FFN<
147│     OutputLayerType,
148│     InitializationRuleType,
149│     MatType
150│ >::Train(const MatType &predictors,
151│          const MatType &responses,
152│          OptimizerType& optimizer,
153│          CallbackTypes&&... callbacks)
154│ {
```

All subsequent functions should also be adjusted accordingly.

**Comment by conradsnicta** [comment]
As a workaround, suggest to provide the arguments to FFN::Train() wrapped in [std::move()](https://en.cppreference.com/w/cpp/utility/move), which will avoid explicit copying of the underlying data.

**Comment by rcurtin** [comment]
This is intentional: we can't pass a const reference, because we may need to shuffle the data.  So `Train()` here is meant to make a copy, or, the user can pass with `std::move()` to avoid the copy if they do not need the data anymore after training.

I noticed that the documentation didn't point this out, so I opened #3490 to improve it.

Thanks for pointing this out!

**Comment by conradsnicta** [comment]
resolved in #3490
